### PR TITLE
Simple cycles interface

### PR DESCRIPTION
**Pull request by Elsee**
Hi! My team took a task from your TODOs list: 
https://github.com/jgrapht/jgrapht/wiki/TODOs#simple-cycles-orgjgraphtalgcycle
We analyze the code and decided that it would be good to make it abstract class instead of interface to avoid code duplication.

### CODE REVIEW

**Comment by d-michail** [comment]
Hi,

the current pattern that we are using is that each algorithm category has a dedicated interface. Interfaces are slowly moving to the package `org.jgrapht.alg.interfaces`. This means that algorithm `PatonCycleBase` which computes cycle bases should have a different interface from the other algorithms which compute all simple cycles. Morever, the naming should follow the algorithms.

We already have a PR for the cycle bases algorithms, thus I would propose to discard the changes in `PatonCycleBase`.

Regarding the other algorithms you should move your interface to `org.jgrapht.alg.interfaces` and call it something like `AllSimpleCyclesAlgorithm`. It should follow the pattern of the other interfaces, such as:

```
public interface AllSimpleCyclesAlgorithm {
   List<GraphPath<V>> findSimpleCycles();
}
```
it should not contain getter-setters for the graph.

Note that we use a 2-version deprecation process which means that the old interface needs to become deprecated and you need to adjust the implementation classes in order to implement both interfaces. If you want to reuse some code, use a package private class which can be extended by all common algorithms.

As far as the implementation is concerned, you implement this like a wrapper which takes the algorithm's result and transforms it. This transformation is already present in the library in the interface `GraphPath` and its implementation called `GraphWalk`. You can simply instantiate a `GraphWalk` using the list of vertices.

Thanks

File: jgrapht-core/src/main/java/org/jgrapht/alg/cycle/DirectedSimpleCycles.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+ * A common interface for classes implementing algorithms for enumeration of the simple cycles of a
+ * directed graph.
+ *
+ * @param <V> the vertex type.
+ * @param <E> the edge type.
+ *
+ * @author Nikolay Ognyanov
+ */
+@Deprecated
+public interface DirectedSimpleCycles<V, E>
```
Comment (d-michail) [inline jgrapht-core/src/main/java/org/jgrapht/alg/cycle/DirectedSimpleCycles.java]: JGraphT uses a 2-release deprecation: 
  - an interfaces becomes deprecated (retaining all methods)
  - next release
  - the interface is deleted
  - next release
  
Also needs `@deprecated` comment in Javadoc with link to the next interface. See other code in the library for examples.

File: jgrapht-core/src/main/java/org/jgrapht/alg/cycle/PatonCycleBase.java [hunk 2]
```diff
@@ -34,8 +39,8 @@
  *
  * @author Nikolay Ognyanov
  */
-public class PatonCycleBase<V, E> extends Cycle<V,E>
-        // was implementing UndirectedCycleBase
+public class PatonCycleBase<V, E>
```
Comment (d-michail) [inline jgrapht-core/src/main/java/org/jgrapht/alg/cycle/PatonCycleBase.java]: We already have a PR about the class `PatonCycleBase` which changes the interface.

File: jgrapht-core/src/main/java/org/jgrapht/alg/cycle/UndirectedCycleBase.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+ * A common interface for classes implementing algorithms for finding a cycle base of an undirected
+ * graph.
+ *
+ * @param <V> the vertex type.
+ * @param <E> the edge type.
+ *
+ * @author Nikolay Ognyanov
+ */
+@Deprecated
+public interface UndirectedCycleBase<V, E>
```
Comment (d-michail) [inline jgrapht-core/src/main/java/org/jgrapht/alg/cycle/UndirectedCycleBase.java]: We already have a PR for cycle bases. No need to do any changes here.

File: jgrapht-core/src/main/java/org/jgrapht/alg/interfaces/AllSimpleCyclesAlgorithm.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+/**
+ * Created by svetl on 26.03.2017.
+ */
+public interface AllSimpleCyclesAlgorithm<V, E> {
+    /**
+     * Returns the graph on which the cycle base search algorithm is executed by this object.
+     *
+     * @return The graph.
+     */
+    Graph<V, E> getGraph();
```
Comment (d-michail) [inline jgrapht-core/src/main/java/org/jgrapht/alg/interfaces/AllSimpleCyclesAlgorithm.java]: Remove this from the new interface. See the other interfaces in the same package as examples.

File: jgrapht-core/src/main/java/org/jgrapht/alg/interfaces/AllSimpleCyclesAlgorithm.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+    /**
+     * Sets the graph on which the cycle base search algorithm is executed by this object.
+     *
+     * @param graph the graph.
+     *
+     * @throws IllegalArgumentException if the argument is <code>null</code>.
+     */
+    void setGraph(Graph<V, E> graph);
```
Comment (d-michail) [inline jgrapht-core/src/main/java/org/jgrapht/alg/interfaces/AllSimpleCyclesAlgorithm.java]: Remove this from the new interface. See the other interfaces in the same package as examples.

File: jgrapht-core/src/main/java/org/jgrapht/alg/interfaces/AllSimpleCyclesAlgorithm.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+    /**
+     * Finds a cycle base of the graph.<br>
```
Comment (d-michail) [inline jgrapht-core/src/main/java/org/jgrapht/alg/interfaces/AllSimpleCyclesAlgorithm.java]: Please remove any mention to "cycle bases". We are dealing with two separate problems here
 - one is the enumeration of all simples cycles of a directed graph
 - the second is finding an undirected cycle bases. 

The new interface is only for the first one. We already have a PR for the second.

File: jgrapht-core/src/main/java/org/jgrapht/alg/interfaces/AllSimpleCyclesAlgorithm.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+     *
+     * @return A list of cycles constituting a cycle base for the graph. Possibly empty but never
+     *         <code>null</code>.
+     *
+     * @throws IllegalArgumentException if the current graph is null.
+     */
+    List<List<V>> findSimpleCycles();
+
+    /**
+     * * Finds a cycle base of the graph.<br>
```
Comment (d-michail) [inline jgrapht-core/src/main/java/org/jgrapht/alg/interfaces/AllSimpleCyclesAlgorithm.java]: Please remove any mention to "cycle bases". We are dealing with two separate problems here

-  one is the enumeration of all simples cycles of a directed graph
-  the second is finding an undirected cycle bases.

The new interface is only for the first one. We already have a PR for the second.

File: jgrapht-core/src/main/java/org/jgrapht/alg/cycle/Cycle.java [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+    /**
+     * * Finds a cycle base of the graph.<br>
+     * Note that the full algorithm is executed on every call since the graph may have changed
+     * between calls.
+     *
+     * @return A list of paths which are cycles. Based on <code> List<List<V>> findCycles() </code>.
+     * @throws IllegalArgumentException
+     */
+    List<GraphPath<V, E>> findCyclePath() throws IllegalArgumentException {
```
Comment (d-michail) [inline jgrapht-core/src/main/java/org/jgrapht/alg/cycle/Cycle.java]: The implementation is already present in class `GraphWalk`. Simply call the method which returns a list of vertices and properly instantiate a `GraphWalk` to return to the users.

**Comment by jkinable** [comment]
@Elsee significant support for cycles was added to JGraphT through PR #381 . Please modify this PR accordingly:

- if this PR contains functionality which is currently missing or if this PR contains implementation improvements, please update this PR wrt the current code base. Alternatively, consider closing this PR.

**Pull request closed without merging**